File: core/src/main/java/com/toocol/termio/core/cache/CredentialCache.java
package com.toocol.termio.core.cache;

import com.toocol.termio.core.auth.core.SshCredential;
import com.toocol.termio.core.term.core.Term;
import com.toocol.termio.utilities.anis.AnisStringBuilder;
import com.toocol.termio.utilities.anis.Printer;
import com.toocol.termio.utilities.functional.Switchable;
import com.toocol.termio.utilities.utils.MessageBox;
import com.toocol.termio.utilities.utils.StrUtil;
import io.vertx.core.json.JsonArray;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/4/2 15:46
 */
public class CredentialCache {

    private static CredentialCache instance;
    private final Set<SshCredential> CREDENTIAL_SET = new TreeSet<>(Comparator.comparingInt(credential -> -1 * credential.getHost().hashCode()));
    private final ReadWriteLock READ_WRITE_LOCK = new ReentrantReadWriteLock();

    private CredentialCache() {

    }

    public synchronized static CredentialCache getInstance() {
        if (instance == null) {
            instance = new CredentialCache();
        }
        return instance;
    }

    public int credentialsSize() {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            return CREDENTIAL_SET.size();
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return 0;
    }

    public String getCredentialsJson() {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            return new JsonArray(new ArrayList<>(CREDENTIAL_SET)).toString();
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return null;
    }

    public void showCredentials() {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            AtomicInteger idx = new AtomicInteger(1);
            Term term = Term.getInstance();
            CREDENTIAL_SET.forEach(credential -> {
                int index = idx.getAndIncrement();
                Printer.print(new AnisStringBuilder()
                        .background(Term.theme.propertiesZoneBgColor)
                        .append(StrUtil.SPACE.repeat(term.getWidth()))
                        .toString()
                );
                term.setCursorPosition(0, term.getCursorPosition()[1]);
                AnisStringBuilder builder = new AnisStringBuilder()
                        .background(Term.theme.propertiesZoneBgColor);
                if (SshSessionCache.getInstance().isAlive(credential.getHost())) {
                    builder.front(Term.theme.sessionAliveColor);
                } else {
                    builder.front(Term.theme.indexFrontColor);
                }
                builder.append("[" + (index < 10 ? "0" + index : index) + "]\t\t").deFront()
                        .front(Term.theme.userHighlightColor).append(credential.getUser()).deFront()
                        .front(Term.theme.atHighlightColor).append("@").deFront()
                        .front(Term.theme.hostHighlightColor).append(credential.getHost()).deFront();
                Printer.println(builder.toString());
            });
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
    }

    public int indexOf(String host, String user) {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            int index = 1;
            for (SshCredential sshCredential : CREDENTIAL_SET) {
                if (sshCredential.getHost().equals(host) && sshCredential.getUser().equals(user)) {
                    return index;
                }
                index++;
            }
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return -1;
    }

    public SshCredential getCredential(int index) {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            int loopIdx = 1;
            for (SshCredential sshCredential : CREDENTIAL_SET) {
                if (loopIdx++ == index) {
                    return sshCredential;
                }
            }
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return null;
    }

    public SshCredential getCredential(String host) {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            for (SshCredential sshCredential : CREDENTIAL_SET) {
                if (sshCredential.getHost().equals(host)) {
                    return sshCredential;
                }
            }
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return null;
    }

    public boolean containsCredential(SshCredential credential) {
        Lock lock = READ_WRITE_LOCK.readLock();
        lock.lock();
        try {
            return CREDENTIAL_SET.contains(credential);
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return false;
    }

    public void addCredential(SshCredential credential) {
        Lock lock = READ_WRITE_LOCK.writeLock();
        lock.lock();
        try {
            CREDENTIAL_SET.add(credential);
        } catch (Exception e) {
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
    }

    public String deleteCredential(int index) {
        Lock lock = READ_WRITE_LOCK.writeLock();
        lock.lock();
        try {
            int tag = 0;
            Iterator<SshCredential> iterator = CREDENTIAL_SET.iterator();
            while (iterator.hasNext()) {
                tag++;
                SshCredential next = iterator.next();
                if (tag == index) {
                    iterator.remove();
                    return next.getHost();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            MessageBox.setExitMessage("Credential operation error.");
            System.exit(-1);
        } finally {
            lock.unlock();
        }
        return null;
    }

    public Collection<Switchable> getAllSwitchable() {
        return new ArrayList<>(CREDENTIAL_SET);
    }
}

File: core/src/main/java/com/toocol/termio/core/cache/CredentialCache.kt
package com.toocol.termio.core.cache

import com.toocol.termio.core.Termio
import com.toocol.termio.core.auth.core.SshCredential
import com.toocol.termio.core.term.core.Term
import com.toocol.termio.utilities.ansi.AnsiStringBuilder
import com.toocol.termio.utilities.ansi.Printer
import com.toocol.termio.utilities.functional.Switchable
import com.toocol.termio.utilities.utils.MessageBox
import com.toocol.termio.utilities.utils.StrUtil
import io.vertx.core.json.JsonArray
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReadWriteLock
import java.util.concurrent.locks.ReentrantReadWriteLock
import java.util.function.Consumer
import kotlin.system.exitProcess

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/4/2 15:46
 */
class CredentialCache private constructor() {
    companion object Instance {
        private val credentialSet: MutableSet<SshCredential?> = TreeSet(Comparator.comparingInt { credential: SshCredential? -> -1 * credential!!.host.hashCode() })
        private val readWriteLock: ReadWriteLock = ReentrantReadWriteLock()
        fun credentialsSize(): Int {
            val lock = readWriteLock.readLock()
            lock.lock()
            try {
                return credentialSet.size
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
        }

        val credentialsJson: String
            get() {
                val lock = readWriteLock.readLock()
                lock.lock()
                try {
                    return JsonArray(ArrayList(credentialSet)).toString()
                } catch (e: Exception) {
                    MessageBox.setExitMessage("Credential operation error.")
                    exitProcess(-1)
                } finally {
                    lock.unlock()
                }
            }

        fun showCredentials() {
            val theme = Term.theme
            val lock = readWriteLock.readLock()
            lock.lock()
            try {
                val idx = AtomicInteger(1)
                val term = Term.instance
                credentialSet.forEach(Consumer { credential: SshCredential? ->
                    val index = idx.getAndIncrement()
                    Printer.print(AnsiStringBuilder()
                        .background(theme.propertiesZoneBgColor.color)
                        .append(StrUtil.SPACE.repeat(Termio.windowWidth))
                        .toString()
                    )
                    term.setCursorPosition(0, term.cursorPosition[1])
                    val builder = AnsiStringBuilder()
                        .background(theme.propertiesZoneBgColor.color)
                    if (SshSessionCache.isAlive(credential!!.host)) {
                        builder.front(theme.sessionAliveColor.color)
                    } else {
                        builder.front(theme.indexFrontColor.color)
                    }
                    builder.append("[" + (if (index < 10) "0$index" else index) + "]\t\t").deFront()
                        .front(theme.userHighlightColor.color).append(credential.user).deFront()
                        .front(theme.atHighlightColor.color).append("@").deFront()
                        .front(theme.hostHighlightColor.color).append(credential.host).deFront()
                    Printer.println(builder.toString())
                })
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
        }

        fun indexOf(host: String, user: String): Int {
            val lock = readWriteLock.readLock()
            lock.lock()
            try {
                var index = 1
                for (sshCredential in credentialSet) {
                    if (sshCredential!!.host == host && sshCredential.user == user) {
                        return index
                    }
                    index++
                }
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
            return -1
        }

        fun getCredential(index: Int): SshCredential? {
            val lock = readWriteLock.readLock()
            lock.lock()
            try {
                var loopIdx = 1
                for (sshCredential in credentialSet) {
                    if (loopIdx++ == index) {
                        return sshCredential
                    }
                }
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
            return null
        }

        fun getCredential(host: String): SshCredential? {
            val lock = readWriteLock.readLock()
            lock.lock()
            try {
                for (sshCredential in credentialSet) {
                    if (sshCredential!!.host == host) {
                        return sshCredential
                    }
                }
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
            return null
        }

        fun containsCredential(credential: SshCredential?): Boolean {
            val lock = readWriteLock.readLock()
            lock.lock()
            try {
                return credentialSet.contains(credential)
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
        }

        fun addCredential(credential: SshCredential?) {
            val lock = readWriteLock.writeLock()
            lock.lock()
            try {
                credentialSet.add(credential)
            } catch (e: Exception) {
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
        }

        fun deleteCredential(index: Int): String? {
            val lock = readWriteLock.writeLock()
            lock.lock()
            try {
                var tag = 0
                val iterator = credentialSet.iterator()
                while (iterator.hasNext()) {
                    tag++
                    val next = iterator.next()
                    if (tag == index) {
                        iterator.remove()
                        return next!!.host
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                MessageBox.setExitMessage("Credential operation error.")
                exitProcess(-1)
            } finally {
                lock.unlock()
            }
            return null
        }

        val allSwitchable: Collection<Switchable>
            get() = ArrayList<Switchable>(credentialSet)
    }
}